src/dataset/FFHQDataset.py

```diff
@@ -0,0 +1,253 @@
+from src.dataset.Dataset import Dataset
+from src.util.FileJar import FileJar
+
+import os, json
+import pandas as pd
+from tqdm import tqdm
+from multiprocessing import Pool, cpu_count
+import numpy as np
+
+
+class FFHQDataset(Dataset):
+    """
+    A subclass to Dataset which provides the ability to extract labels
+    from the dataset. These labels are saved to disk which allows subsequent
+    creations of the dataset to be performed without extraction the labels
+    again.
+
+    Before creating a FFQDataset, a directory should
+    exist with the name of "FFHQ_`resolution`" in the dataset directory.
+    There should also exist appriopriate directories for the labels ("label/")
+    and images ("image/").
+    """
+
+    # Dataset name
+    DS_NAME = "FFHQ"
+
+    # Subdirectory name to the images
+    DS_DIR_IMAGES = "image/"
+
+    # Subdirectory name to the labels
+    DS_DIR_LABELS = "label/"
+
+    # Filename of the extracted labels
+    DS_LABELS_NAME = "labels.pkl"
+
+    def __init__(self, resolution: int = 256):
+        """
+        Constructor which extracts labels from the dataset if not prevously done
+        in the super class with the help of `init_files()`,
+        this process is managed by a `FileJar`. The name of the dataset is a
+        combination of FFHQ and the specified resolution.
+
+        Args:
+            resolution (int, optional): The native resolution of the dataset.
+                Defaults to 256.
+        """
+        name = self.DS_NAME + "_" + str(resolution)
+        super(FFHQDataset, self).__init__(name)
+        self._resolution = resolution
+
+    def get_resolution(self) -> int:
+        """
+        Returns the resolution of the dataset.
+
+        Returns:
+            int: the resolution of the dataset.
+        """
+        return self._resolution
+
+    def init_files(self) -> tuple[FileJar, pd.DataFrame]:
+        """
+        Initilizes the label files related to the dataset.
+        Extract labels from the dataset and saves them with a
+        FileJar as well as a pd.DataFrame.
+        Note that the use of FileJar allows for saving the dataframe such
+        that it can be loaded again next time without any extraction needed.
+
+        Returns:
+            tuple[FileJar, pd.DataFrame]: FileJar with save paths and
+                a pd.DataFrame containing labels.
+        """
+        file_jar = FileJar(self._ds_dir)
+        df = file_jar.get_file(self.DS_LABELS_NAME, pd.read_pickle)
+        if df is None:
+            return self._parse_json()
+        else:
+            print(f"======= Labels already extracted, skipping extraction =======")
+            return file_jar, df
+
+    def _parse_json(self) -> tuple[FileJar, pd.DataFrame]:
+        """
+        Extract labels from the dataset and saves them with a
+        FileJar as well as a pd.DataFrame.
+        Extraction is done by multiprocessing the tasks,
```

Comment:
If a new paragraph is desired, there has to be a whole blank line between I believe (I guess that this is why "Extraction..." starts on a new line).